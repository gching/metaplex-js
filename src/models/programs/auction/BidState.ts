import BN from 'bn.js';
import { PublicKey } from '@solana/web3.js';
import { Bid } from './Bid';
import { BorshSchema } from '../../utils/types';

export enum BidStateType {
  EnglishAuction = 0,
  OpenEdition = 1,
}

export class BidState {
  static SCHEMA: BorshSchema = new Map([
    [
      BidState,
      {
        kind: 'struct',
        fields: [
          ['type', 'u8'],
          ['bids', [Bid]],
          ['max', 'u64'],
        ],
      },
    ],
    ...Bid.SCHEMA,
  ]);

  type: BidStateType;
  bids: Bid[];
  max: BN;

  public getWinnerAt(winnerIndex: number): PublicKey | null {
    const convertedIndex = this.bids.length - winnerIndex - 1;

    if (convertedIndex >= 0 && convertedIndex < this.bids.length) {
      return this.bids[convertedIndex].key;
    } else {
      return null;
    }
  }

  public getWinnerIndex(bidder: PublicKey): number | null {
    if (!this.bids) return null;

    const index = this.bids.findIndex(
      (b) => b.key.toBase58() === bidder.toBase58(),
    );
    // auction stores data in reverse order
    if (index !== -1) {
      const zeroBased = this.bids.length - index - 1;
      return zeroBased < this.max.toNumber() ? zeroBased : null;
    } else return null;
  }

  constructor(args: { type: BidStateType; bids: Bid[]; max: BN }) {
    this.type = args.type;
    this.bids = args.bids;
    this.max = args.max;
  }
}
