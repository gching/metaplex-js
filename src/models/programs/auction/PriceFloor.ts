import BN from 'bn.js';
import { BorshSchema } from '../../utils/types';

export enum PriceFloorType {
  None = 0,
  Minimum = 1,
  BlindedPrice = 2,
}

export class PriceFloor {
  static SCHEMA: BorshSchema = new Map([
    [
      PriceFloor,
      {
        kind: 'struct',
        fields: [
          ['type', 'u8'],
          ['hash', [32]],
        ],
      },
    ],
  ]);

  type: PriceFloorType;
  // It's an array of 32 u8s, when minimum, only first 8 are used (a u64), when blinded price, the entire
  // thing is a hash and not actually a public key, and none is all zeroes
  hash: Uint8Array;

  minPrice?: BN;

  constructor(args: {
    type: PriceFloorType;
    hash?: Uint8Array;
    minPrice?: BN;
  }) {
    this.type = args.type;
    this.hash = args.hash || new Uint8Array(32);
    if (this.type === PriceFloorType.Minimum) {
      if (args.minPrice) {
        this.hash.set(args.minPrice.toArrayLike(Buffer, 'le', 8), 0);
      } else {
        this.minPrice = new BN(
          (args.hash || new Uint8Array(0)).slice(0, 8),
          'le',
        );
      }
    }
  }
}
