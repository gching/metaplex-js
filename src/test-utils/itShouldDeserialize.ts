import { PublicKey } from '@solana/web3.js';
import { serialize } from 'borsh';
import { BorshSchema } from '../models/utils/types';

export function itShouldDeserialize<E>(
  cls: {
    SCHEMA: BorshSchema;
    // eslint-disable-next-line @typescript-eslint/ban-types
    decode: Function;
  },
  received: E,
  accountPubKey = PublicKey.default,
): void {
  const data = Buffer.from(serialize(cls.SCHEMA, received));

  const deserialized = cls.decode({
    pubkey: accountPubKey,
    account: {
      data,
    },
  }) as E;

  expect(deserialized).toDeepEqual(received);
}
