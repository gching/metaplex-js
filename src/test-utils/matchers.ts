import { isEqualWith } from 'lodash';
import BN from 'bn.js';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toDeepEqual: <E = any>(expected: E) => R;
    }
  }
}

expect.extend({
  toDeepEqual<E = any>(received: E, expected: E) {
    const pass = isEqualWith(received, expected, (objValue, otherValue) => {
      if (BN.isBN(objValue) && BN.isBN(otherValue)) {
        return objValue.eq(otherValue);
      }
      return undefined;
    });

    if (pass) {
      return {
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        message: () => `expected ${received} not to deep equal ${expected}`,
        pass: true,
      };
    }

    return {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      message: () => `expected ${received} to deep equal ${expected}`,
      pass: false,
    };
  },
});

export default undefined;
