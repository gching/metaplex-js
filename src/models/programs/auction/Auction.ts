import { deserializeUnchecked } from 'borsh';
import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { BorshSchema } from '../../utils/types';
import { PriceFloor } from './PriceFloor';
import { BidState } from './BidState';
import { AccountWithPublicKey } from '../../../utils/privateTypes';

export enum AuctionState {
  Created = 0,
  Started,
  Ended,
}

export class Auction {
  static SCHEMA: BorshSchema = new Map([
    [
      Auction,
      {
        kind: 'struct',
        fields: [
          ['authority', 'pubkey'],
          ['tokenMint', 'pubkey'],
          ['lastBid', { kind: 'option', type: 'u64' }],
          ['endedAt', { kind: 'option', type: 'u64' }],
          ['endAuctionAt', { kind: 'option', type: 'u64' }],
          ['auctionGap', { kind: 'option', type: 'u64' }],
          ['priceFloor', PriceFloor],
          ['state', 'u8'],
          ['bidState', BidState],
          ['totalUncancelledBids', { kind: 'option', type: 'u64' }],
        ],
      },
    ],
    ...PriceFloor.SCHEMA,
    ...BidState.SCHEMA,
  ]);

  accountPublicKey?: PublicKey;
  /// Pubkey of the authority with permission to modify this auction.
  authority: PublicKey;
  /// Token mint for the SPL token being used to bid
  tokenMint: PublicKey;
  /// The time the last bid was placed, used to keep track of auction timing.
  lastBid: BN | undefined;
  /// Slot time the auction was officially ended by.
  endedAt: BN | undefined;
  /// End time is the cut-off point that the auction is forced to end by.
  endAuctionAt: BN | undefined;
  /// Gap time is the amount of time in slots after the previous bid at which the auction ends.
  auctionGap: BN | undefined;
  /// Minimum price for any bid to meet.
  priceFloor: PriceFloor;
  /// The state the auction is in, whether it has started or ended.
  state: AuctionState;
  /// Auction Bids, each user may have one bid open at a time.
  bidState: BidState;
  /// Total uncancelled bids
  totalUncancelledBids: BN;
  /// Used for precalculation on the front end, not a backend key
  bidRedemptionKey?: PublicKey;

  constructor(args: {
    accountPublicKey?: PublicKey;
    authority: PublicKey;
    tokenMint: PublicKey;
    lastBid?: BN;
    endedAt?: BN;
    endAuctionAt?: BN;
    auctionGap?: BN;
    priceFloor: PriceFloor;
    state: AuctionState;
    bidState: BidState;
    totalUncancelledBids: BN;
  }) {
    this.accountPublicKey = args.accountPublicKey;
    this.authority = args.authority;
    this.tokenMint = args.tokenMint;
    this.lastBid = args.lastBid;
    this.endedAt = args.endedAt;
    this.endAuctionAt = args.endAuctionAt;
    this.auctionGap = args.auctionGap;
    this.priceFloor = args.priceFloor;
    this.state = args.state;
    this.bidState = args.bidState;
    this.totalUncancelledBids = args.totalUncancelledBids;
  }

  static decode(accountWithPubKey: AccountWithPublicKey): Auction {
    const auction = deserializeUnchecked(
      Auction.SCHEMA,
      Auction,
      accountWithPubKey.account.data,
    ) as Auction;

    return new Auction({
      ...auction,
      accountPublicKey: accountWithPubKey.pubkey,
    });
  }
}
