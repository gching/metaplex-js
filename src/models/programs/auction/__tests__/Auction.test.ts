import { Auction, AuctionState } from '../Auction';
import BN from 'bn.js';
import { PublicKey } from '@solana/web3.js';
import { BidState, BidStateType } from '../BidState';
import { PriceFloor, PriceFloorType } from '../PriceFloor';
import { itShouldDeserialize } from '../../../../test-utils/itShouldDeserialize';

describe('Auction', () => {
  it('should instantiate with the correct values', () => {
    const auction = new Auction({
      authority: PublicKey.default,
      tokenMint: PublicKey.default,
      totalUncancelledBids: new BN(0),
      bidState: new BidState({
        bids: [],
        max: new BN(0),
        type: BidStateType.EnglishAuction,
      }),
      priceFloor: new PriceFloor({
        type: PriceFloorType.BlindedPrice,
      }),
      state: AuctionState.Created,
    });

    expect(auction).toMatchSnapshot();
  });

  it('should deserialize with the correct values', () => {
    const auction = new Auction({
      authority: PublicKey.default,
      lastBid: new BN(0),
      tokenMint: PublicKey.default,
      totalUncancelledBids: new BN(0),
      bidState: new BidState({
        bids: [],
        max: new BN(0),
        type: BidStateType.EnglishAuction,
      }),
      priceFloor: new PriceFloor({
        type: PriceFloorType.BlindedPrice,
      }),
      state: AuctionState.Created,
    });

    auction.accountPublicKey = PublicKey.default;

    itShouldDeserialize(Auction, auction);
  });
});
