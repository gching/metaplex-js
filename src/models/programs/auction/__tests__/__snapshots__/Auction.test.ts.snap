// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auction should instantiate with the correct values 1`] = `
Auction {
  "accountPublicKey": undefined,
  "auctionGap": undefined,
  "authority": PublicKey {
    "_bn": "00",
  },
  "bidState": BidState {
    "bids": Array [],
    "max": "00",
    "type": 0,
  },
  "endAuctionAt": undefined,
  "endedAt": undefined,
  "lastBid": undefined,
  "priceFloor": PriceFloor {
    "hash": Uint8Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    "type": 2,
  },
  "state": 0,
  "tokenMint": PublicKey {
    "_bn": "00",
  },
  "totalUncancelledBids": "00",
}
`;
